### ISSUE TITLE
Safe to increasing bucketing on heavily used dedicated production OpenTSDB cluster?

### ISSUE THREAD

**Issue by HariSekhon** [OpenTSDB/opentsdb#1244]
Hello OpenTSDB gurus!

I need to confirm whether it's safe to increase the `tsd.storage.salt.buckets` setting across all the TSDs without losing data?

I know you can't change salt width on an existing cluster as it would ruin the data alignment and would also be pointless (as per our discussion in #874 a couple of years back), but I've re-checked the docs and can't see explicitly whether it's safe or unsafe to change the bucketing when there is existing data we need to preserve. I believe it is safe to change the bucket size, although suspect some TSDs / Grafana will miss data in results until all TSDs have the same bucket size? I expect their query results will be complete once all TSDs pick up the updated config and that no writes will be lost either as they'll be sent to new buckets in new pre-split regions.

## Interesting OpenTSDB cluster case study problem prompting this question

I've inherited a heavily used OpenTSDB 2.3 cluster with 250,000+ metrics (10TB) running on HBase 1.1.2 on Hortonworks HDP 2.6 - this cluster has good hardware and is dedicated to OpenTSDB + HBase - there are no analytics jobs and there are no other HBase tables.

It started with 5 nodes and was configured with 5 buckets and pre-splitting along \x00, \x01, \x02, \x03, \x04 key boundaries with sub-splits at the next byte too, but has now grown to 8 nodes and I'm noticing significant skew in usage of the RegionServers and regions, despite the region counts being 140 across all RegionServers, some RegionServers are doing 30,000 requests per second and others doing only 10-15,000 requests per second, and I've also noticed signficant skew in request counts between regions within each single RegionServer as well.

They have been suffering from intermittent call queue full exceptions in Grafana as well as occasional YouAreDeadException due to large G1GC pauses causing ZooKeeper znode loss and HMaster to declare RS dead and reassign regions, even though G1GC target is set to 100 ms with RS heap = 25GB and handler count = 100. Usually one RegionServer gets overloaded once in a while over the course of a few days, then for the next few days it affects a different RegionServer, and so on like a rolling issue.

I believe it may be beneficial to increase the bucketing to 8 buckets with 3 extra manual splits and migrations to the least loaded RegionServers as 3 of them seem less loaded than 5 others., likely due to the original 5 buckets still being in place.

It looks like HBase still doesn't use useful heuristics around region load when considering region migrations, with no anti-affinity of heavily used regions to prevent them ending up on the same RS and causing this RegionServer usage skew. I suspect under-utilised pre sub-splits are actually making this worse by allowing some RegionServers to host multiple lightly used regions while others host multiple heavily used regions, because HBase is only considering region counts, somewhat undermining the bucketing, so I'm not sure how beneficial the extra bucketing will be. I also checked uid cardinality which seems fine with uid row count at 5.6M, tagk just over 1600, and tagv at 3.2M as reported by both hbase rows scans and TSDB stats page.

I suspect that a higher degree of bucketing will improve spread more evenly across regions more so than relying on random uids being spread across sub-split regions and therefore give more opportunity for better balance of load across RegionServers.

Is it safe to change bucketing? Do you think it will help or should I just concentrate on manual region rebalance. The problem I have with that is that I don't see it as a good long term solution as this problem may recur.

**Comment by johnewing1** [comment]
```
It looks like HBase still doesn't use useful heuristics around region load when considering region migrations,
```
This isn't strictly true the stochastic load balancer can consider read and write load when moving regions across the cluster. Until recently it performed so badly it that it didn't work effectively

The work by these devs has drastically improved it and the changes are in new versions of hbase
https://www.kahlil.me/hbase-balancer/

You might already be using it, but if not the hbase region inspector makes balancing regions for read/ write load with drag and drop quite easy compare with doing cli move commands
https://github.com/kakao/hbase-region-inspector

```
 some RegionServers are doing 30,000 requests per second and others doing only 10-15,000 requests per second

**Comment by HariSekhon** [comment]
Yes the HBase balancer is much better than years gone by when it only considered region counts and for the most part it does an ok job of balancing 4-5 different heuristics, although I had to force balance by table to fix the table region skew and I don't think it can help enough given our old bucketing setup which was configured when we only had 5 TSDB + HBase servers.

Regardless of how good the balancer gets though I believe that 5 buckets for > 5 servers will not give as good a spread as it could be receiving so I'd like to increase the bucketing, although I want to confrim with one of the devs who has a deep understanding of the TSD bucketing that this is safe to do from a historical data perspective.

**Comment by manolama** [comment]
The buckets is a balance in that you want, generally, some percentage of the region servers you'll have in production but not more than 20 or so since that many parallel queries must be fired for every metric query. The more you have the more outstanding queries there will be with contention. So for 5 servers, 5 buckets is fine. But if you expect to ever hit 10 servers or 20 or hundreds, then just set it to 16 or 20.
